### PR TITLE
:sparkles: Added filter function to controller reconciler

### LINKED ISSUE #7949: Support passing custom predicates options to KubeadmConfigReconciler
**User Story**

We are importing cluster-api controllers in [kurator](https://github.com/kurator-dev/kurator), basically for installing k8s on any platform, especially on running VMs.

In our case, on developing installation on VMs, we depend on kubeadmControlplane API to provide k8s controlplane configurations as well.  But we has the kubeadmControlplane has reference to kurator customMachine API.  So there are lots of errors from KubeadmConfigReconciler.

The basic idea is can we add custom filter function in `controller.Options`, so we can pass the filter function as below. 

```
	c, err := ctrl.NewControllerManagedBy(mgr).
		For(&controlplanev1.KubeadmControlPlane{}, builder.WithPredicates(...) ).
		Owns(&clusterv1.Machine{}).
		WithOptions(options).
		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue)).
		Build(r)
```

**Detailed Description**

[A clear and concise description of what you want to happen.]

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7949

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
@fabriziopandini @sbueringer 

What does the apidiff incompatible mean? How could I make the CI happy?

```
sigs.k8s.io/cluster-api/controllers
  Incompatible changes:
  - ClusterClassReconciler: old is comparable, new is not
```

File: bootstrap/kubeadm/controllers/alias.go [hunk 1]
```diff
@@ -42,6 +43,9 @@ type KubeadmConfigReconciler struct {
 	// WatchFilterValue is the label value used to filter events prior to reconciliation.
 	WatchFilterValue string
 
+	// Filter checks if a KubeadmConfig object should be reconciled.
+	Filter predicates.Filter
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/controllers/alias.go:47]: /hold

My personal opinion is that we should not add a new knob to filter issues we reconcile which is orthogonal to WatchFilterValue and its possible improvements being discussed in https://github.com/kubernetes-sigs/cluster-api/issues/7775. 
IMO those efforts should converge and come up with a common definition of what a filter is, that could translate into e.g. a watch filter struct or other solutions.

Also, worth to notice that such type of change in order to actually work should be accepted by provider implementers as well, and this requires a discussion during office hours or a small proposal to rally on async.

File: bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go [hunk 2]
```diff
@@ -160,6 +162,9 @@ func (r *KubeadmConfigReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		log.Error(err, "Failed to get config")
 		return ctrl.Result{}, err
 	}
+	if r.Filter != nil && !r.Filter.Filter(config) {
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:165]: Filters should be applied before actually entering in the reconcile loop (e.g. in while setting up controllers)

File: bootstrap/kubeadm/controllers/alias.go (hunk 1 above)
Comment (hzxuzhonghu) [reply bootstrap/kubeadm/controllers/alias.go:47]: I am ok with moving discussion to  #7775.

File: bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go (hunk 2 above)
Comment (hzxuzhonghu) [reply bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:165]: But then how can we filter out resources that are triggered by another kind of resource?

**Pull request closed without merging**